test_transition_fs_sync_cow_full fails on fedora 35

###### Arch: amd64
###### Distributive: Fedora 35
###### Filesystem: ext4 (no RAID, no LVM)
###### Device type: qcow2
###### Details:

```
test_destroy_active_incremental (test_destroy.TestDestroy) ... ok
test_destroy_active_snapshot (test_destroy.TestDestroy) ... ok
test_destroy_dormant_incremental (test_destroy.TestDestroy) ... ok
test_destroy_dormant_snapshot (test_destroy.TestDestroy) ... ok
test_destroy_nonexistent_device (test_destroy.TestDestroy) ... ok
test_destroy_unverified_incremental (test_destroy.TestDestroy) ... ok
test_destroy_unverified_snapshot (test_destroy.TestDestroy) ... ok
test_multipart_modify_origins (test_multipart.TestMultipart) ... skipped 'Multipart testcase requires disk, not a partition.'
test_multipart_setup_volumes_same_disk (test_multipart.TestMultipart) ... skipped 'Multipart testcase requires disk, not a partition.'
test_multipart_setup_volumes_write_last_after_destroy (test_multipart.TestMultipart) ... skipped 'Multipart testcase requires disk, not a partition.'
test_setup_2_volumes (test_setup.TestSetup) ... ok
test_setup_already_tracked_volume (test_setup.TestSetup) ... ok
test_setup_cow_file_path_is_dir (test_setup.TestSetup) ... ok
test_setup_invalid_minor (test_setup.TestSetup) ... ok
test_setup_readonly_volume (test_setup.TestSetup) ... ok
test_setup_unmounted_volume (test_setup.TestSetup) ... ok
test_setup_volume (test_setup.TestSetup) ... ok
test_setup_volume_path_is_dir (test_setup.TestSetup) ... ok
test_modify_origin (test_snapshot.TestSnapshot) ... ok
test_next_available_minor (test_snapshot.TestSnapshot) ... ok
test_track_writes (test_snapshot.TestSnapshot) ... ok
test_transition_active_incremental (test_transition_incremental.TestTransitionToIncremental) ... ok
test_transition_active_snapshot (test_transition_incremental.TestTransitionToIncremental) ... ok
test_transition_fs_sync_cow_full (test_transition_incremental.TestTransitionToIncremental) ... FAIL
test_transition_mod_sync_cow_full (test_transition_incremental.TestTransitionToIncremental) ... skipped 'Kernel flushed before module'
test_transition_nonexistent_snapshot (test_transition_incremental.TestTransitionToIncremental) ... ok

======================================================================
FAIL: test_transition_fs_sync_cow_full (test_transition_incremental.TestTransitionToIncremental)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elastio/elastio-snap/tests/test_transition_incremental.py", line 58, in test_transition_fs_sync_cow_full
    self.assertTrue(err == errno.EINVAL or err == errno.EFBIG)
AssertionError: False is not true
```

--- a/src/elastio-snap.c
+++ b/src/elastio-snap.c
@@ -3349,10 +3349,9 @@ error:
 static inline void wait_for_bio_complete(struct snap_device *dev)
 {
 	struct bio_queue *bq = &dev->sd_cow_bios;
-	if (atomic64_read(&dev->sd_received_cnt) != atomic64_read(&dev->sd_processed_cnt))
-		wait_event_interruptible_timeout(bq->event,
-				atomic64_read(&dev->sd_submitted_cnt) == atomic64_read(&dev->sd_processed_cnt),
-				msecs_to_jiffies(500));
+	wait_event_interruptible_timeout(bq->event,
+			atomic64_read(&dev->sd_submitted_cnt) == atomic64_read(&dev->sd_processed_cnt),
+			msecs_to_jiffies(500));
 }
 
 #ifdef HAVE_BIO_ENDIO_INT
@@ -4930,7 +4929,6 @@ static int ioctl_elastio_snap_info(struct elastio_snap_info *info){
 	if(ret) goto error;
 
 	dev = snap_devices[info->minor];
-	wait_for_bio_complete(dev);
 
 	tracer_elastio_snap_info(dev, info);
 

--- a/src/elastio-snap.c
+++ b/src/elastio-snap.c
@@ -1111,6 +1111,7 @@ static void elastio_snap_proc_stop(struct seq_file *m, void *v);
 static int elastio_snap_proc_open(struct inode *inode, struct file *filp);
 static int elastio_snap_proc_release(struct inode *inode, struct file *file);
 
+#define WAIT_SUBMITTED_BIOS_MSEC 500
 // wait msec value to be at least 100 msec as wait loop uses it by msleep of (100) timeout pieces
 #define ELASTIO_SNAP_WAIT_FOR_RELEASE_MSEC              500
 #define ELASTIO_SNAP_WAIT_FOR_RELEASE_MAX_SLEEP_COUNT   100
@@ -3346,12 +3347,16 @@ error:
 	bio_free_clone(bio);
 }
 
+/** Resolves issue https://github.com/elastio/elastio-snap/issues/170 */
 static inline void wait_for_bio_complete(struct snap_device *dev)
 {
 	struct bio_queue *bq = &dev->sd_cow_bios;
-	wait_event_interruptible_timeout(bq->event,
+	if (!wait_event_interruptible_timeout(bq->event,
 			atomic64_read(&dev->sd_submitted_cnt) == atomic64_read(&dev->sd_processed_cnt),
-			msecs_to_jiffies(500));
+			msecs_to_jiffies(WAIT_SUBMITTED_BIOS_MSEC))) {
+		LOG_DEBUG("failed wait for all submitted BIOs to be processed after %d ms. bio submitted = %lld, bio processed = %lld",
+				WAIT_SUBMITTED_BIOS_MSEC, atomic64_read(&dev->sd_submitted_cnt), atomic64_read(&dev->sd_processed_cnt));
+	}
 }
 
 #ifdef HAVE_BIO_ENDIO_INT

--- a/src/elastio-snap.c
+++ b/src/elastio-snap.c
@@ -3354,7 +3354,7 @@ static inline void wait_for_bio_complete(struct snap_device *dev)
 	if (!wait_event_interruptible_timeout(bq->event,
 			atomic64_read(&dev->sd_submitted_cnt) == atomic64_read(&dev->sd_processed_cnt),
 			msecs_to_jiffies(WAIT_SUBMITTED_BIOS_MSEC))) {
-		LOG_DEBUG("failed wait for all submitted BIOs to be processed after %d ms. bio submitted = %lld, bio processed = %lld",
+		LOG_WARN("failed wait for all submitted BIOs to be processed after %d ms. bio submitted = %lld, bio processed = %lld",
 				WAIT_SUBMITTED_BIOS_MSEC, atomic64_read(&dev->sd_submitted_cnt), atomic64_read(&dev->sd_processed_cnt));
 	}
 }

--- a/src/elastio-snap.c
+++ b/src/elastio-snap.c
@@ -1077,6 +1077,7 @@ struct snap_device{
 #endif
 	atomic64_t sd_submitted_cnt; //count of read clones submitted to underlying driver
 	atomic64_t sd_received_cnt; //count of read clones submitted to underlying driver
+	atomic64_t sd_processed_cnt; //count of read clones processed in snap_cow_thread()
 };
 
 static long ctrl_ioctl(struct file *filp, unsigned int cmd, unsigned long arg);
@@ -3224,6 +3225,7 @@ static int snap_cow_thread(void *data){
 				}
 			}
 
+			atomic64_inc(&dev->sd_processed_cnt);
 			bio_free_clone(bio);
 		}
 	}
@@ -3342,6 +3344,15 @@ error:
 	tracer_set_fail_state(dev, ret);
 	tp_put(tp);
 	bio_free_clone(bio);
+}
+
+static inline void wait_for_bio_complete(struct snap_device *dev)
+{
+	struct bio_queue *bq = &dev->sd_cow_bios;
+	if (atomic64_read(&dev->sd_received_cnt) != atomic64_read(&dev->sd_processed_cnt))
+		wait_event_interruptible_timeout(bq->event,
+				atomic64_read(&dev->sd_submitted_cnt) == atomic64_read(&dev->sd_processed_cnt),
+				msecs_to_jiffies(500));
 }
 
 #ifdef HAVE_BIO_ENDIO_INT
@@ -4251,6 +4262,7 @@ static int __tracer_setup_snap(struct snap_device *dev, unsigned int minor, stru
 
 	atomic64_set(&dev->sd_submitted_cnt, 0);
 	atomic64_set(&dev->sd_received_cnt, 0);
+	atomic64_set(&dev->sd_processed_cnt, 0);
 
 	return 0;
 
@@ -4793,6 +4805,8 @@ static int ioctl_destroy(unsigned int minor){
 	}
 
 	dev = snap_devices[minor];
+	wait_for_bio_complete(dev);
+
 	tracer_destroy(snap_devices[minor]);
 	kfree(dev);
 
@@ -4810,6 +4824,7 @@ static int ioctl_transition_inc(unsigned int minor){
 	if(ret) goto error;
 
 	dev = snap_devices[minor];
+	wait_for_bio_complete(dev);
 
 	//check that the device is not in the fail state
 	ret = tracer_read_fail_state(dev);
@@ -4849,6 +4864,7 @@ static int ioctl_transition_snap(unsigned int minor, const char *cow_path, unsig
 	if(ret) goto error;
 
 	dev = snap_devices[minor];
+	wait_for_bio_complete(dev);
 
 	//check that the device is not in the fail state
 	if(tracer_read_fail_state(dev)){
@@ -4885,6 +4901,7 @@ static int ioctl_reconfigure(unsigned int minor, unsigned long cache_size){
 	if(ret) goto error;
 
 	dev = snap_devices[minor];
+	wait_for_bio_complete(dev);
 
 	//check that the device is not in the fail state
 	if(tracer_read_fail_state(dev)){
@@ -4913,6 +4930,7 @@ static int ioctl_elastio_snap_info(struct elastio_snap_info *info){
 	if(ret) goto error;
 
 	dev = snap_devices[info->minor];
+	wait_for_bio_complete(dev);
 
 	tracer_elastio_snap_info(dev, info);
 

